src/main/java/com/salesforce/perfeng/uiperf/imageoptimization/service/ImageOptimizationService.java

```diff
@@ -0,0 +1,1234 @@
+            try {
+                IOUtils.copy(is, writer, StandardCharsets.UTF_8);
+                final StringBuilder errorMessage = new StringBuilder("Optimization failed with edit code: ").append(ps.exitValue()).append(". ").append(writer);
+                if (ps.exitValue() == 127 /* command not found */) {
+                    throw new ThirdPartyBinaryNotFoundException(binaryApplicationName, "Most likely this is due to required libraries not being installed on the OS. On Ubuntu run \"sudo apt-get install libjpeg62:i386\".", new RuntimeException(errorMessage.toString()));
+                }
+                throw ImageFileOptimizationException.getInstance(originalFile, new RuntimeException(errorMessage.toString()));
+            } catch (final IOException ioe) {
+                logger.error("Unable to redirect error output for child process for " + originalFile, ioe);
+            }
+        } catch(final ThirdPartyBinaryNotFoundException | ImageFileOptimizationException ifoe) {
+            throw ifoe;
+        } catch(final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     *
+     * Optimizes all of the passed in images. This process is multi-threaded so
+     * that the number of threads is equal to the number of CPUs.
+     *
+     * @param conversionType If and how to handle converting images from one
+     *                       type to another.
+     * @param includeWebPConversion If <code>true</code> then the a WebP version
+     *                              of the image will also be generated (if it
+     *                              is smaller).
+     * @param files The images to optimize
+     * @return The results from the optimization. All items in the {@link List}
+     *         are considered optimized, not <code>null</code>, and will exclude
+     *         images that could not be optimized to a smaller size.
+     * @throws ImageFileOptimizationException If there are any issues optimizing
+     *                                        an image.
+     * @throws TimeoutException Happens if it takes to long to optimize an
+     *                          image.
+     * @see #optimizeAllImages(com.salesforce.perfeng.uiperf.imageoptimization.service.IImageOptimizationService.FileTypeConversion, boolean, File...)
+     * @see com.salesforce.perfeng.uiperf.imageoptimization.service.IImageOptimizationService#optimizeAllImages(FileTypeConversion, boolean, Collection)
+     */
+    @Override
+    public List<OptimizationResult<C>> optimizeAllImages(final FileTypeConversion conversionType, final boolean includeWebPConversion, final Collection<File> files) throws ImageFileOptimizationException, TimeoutException {
+        if ((files == null) || files.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        final CompletionService<OptimizationResult<C>> completionService = new ExecutorCompletionService<>(executorService);
+
+        int i = 0;
```

Comment:
Hmm... is github just being annoying and not showing me the usage of i?

Reply:
(looks like it is not used anywhere)

Reply:
It is used on line 524